R/dependencies.R

```diff
@@ -0,0 +1,410 @@
+    if (vertex$name %in% reports) {
+      reports <- setdiff(reports, vertex$name)
+    }
+    reports <- c(reports, vertex$name) ## add it to the list
+  }
+
+  if (length(vertex$children) > 0) {
+    for (i in 1:length(vertex$children)) {
+      child <- vertex$children[[i]]
+
+      reports <- out_ot_date_reports(child, reports)
+    }
+  }
+  reports
+}
+
+##' @title Print the dependency tree for a given report using orderly log
+##'
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param con A connection to a database
+##' @param root
+##' @param locate
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##' @param propagate A boolean indicating if we want to propogate out of date
+##'                  through the tree
+##'
+##' @export
+orderly_build_dep_tree <- function(name, id = "latest", root = NULL,
+                                   locate = TRUE, upstream = FALSE, con = NULL,
+                                   propagate = FALSE, max_depth = 100,
+                                   list_all = FALSE) {
+  if (is.null(con)) {
+    con <- orderly_db("destination", orderly_config_get(root, locate))
+    on.exit(DBI::dbDisconnect(con))
+  }
+
+  dep_tree <- build_tree(name = name, id = id, depth = max_depth, con = con,
+                         upstream = upstream, list_all = list_all)
+
+  # propagate out-of-date
+  if (propagate) {
+    propagate(dep_tree$root, upstream)
+  }
+
+  dep_tree
+}
+
+propagate <- function(vertex, upstream) {
+  for (child in vertex$children) {
+    if (!upstream) {
+      if (vertex$out_of_date) {
+        child$out_of_date = TRUE
+      }
+    }
+
+    propagate(child, upstream)
+
+    if (upstream) {
+      if (child$out_of_date) {
+        vertex$out_of_date = TRUE
+      }
+    }
+  }
+}
+
+Vertex <- R6::R6Class("Vertex", list(
+  parent = NULL,
+  children = list(),
+  name = NULL,
+  id = NULL,
+  out_of_date = NULL,
+  initialize = function(parent, name, id, out_of_date) {
+    # add some type safety here
+    self$parent <- parent
+    self$name <- name
+    self$id <- id
+    self$out_of_date <- out_of_date
+  },
+  add_child = function(child) {
+    self$children <- append(self$children, list(child))
+  },
+  format = function() {
+    return(sprintf("%s [%s]", self$name, self$id))
+  }
+  )
+)
+
+
+Tree <- R6::R6Class("Tree", list(
+  vertices = NULL,
+  root = NULL,
+  message = NULL,
+  initialize = function(root) {
+    self$vertices <- append(self$vertices, list(root))
+    self$root <- root
+    self$message <- NULL
+  },
+  add_child = function(parent, name, id, out_of_date) {
+    child <- Vertex$new(parent, name, id, out_of_date)
+    parent$add_child(child)
+    self$vertices <- append(self$vertices, list(child))
+    return(child)
+  },
+  # this is stupidly hacky to get the formatting right
+  format_helper = function(vertex = self$root, fvector = c(), str="") {
```

Comment:
this should be a private method, and we should think about making it not stupidly hacky, or at least explaining what the problem is